### ISSUE TITLE
issue when deploying to Heroku

### ISSUE THREAD

**Issue by acandael** [Aupajo/middleman-search_engine_sitemap#2]
I just implemented the middleman-search_engine_sitemap gem to my Middleman project and it works fine locally.

However when I push my changes to Heroku, I get this build error:

    error  tmp/sitemap.xml
    There were errors during this build, re-run with `middleman build --verbose` to see the full    exception.
   rake aborted!
   Command failed with status (1): [middleman build...]
   /tmp/build_b3d8907b-d401-4248-91c2-027fd6b69c1d/Rakefile:3:in `block (2 levels) in <top (required)>'

    !
    !     Precompiling assets failed.
    ! 

    !     Push rejected, failed to compile Ruby app

     To git@heroku.com:morning-inlet-7066.git
      ! [remote rejected] master -> master (pre-receive hook declined)
      error: failed to push some refs to 'git@heroku.com:morning-inlet-7066.git'

Does anyone have a clue what the problem might be?
My repository is at:
https://github.com/acandael/personal-site

thanks for your help,

Anthony

**Comment by subshad** [comment]
@Aupajo Not intending to resurrect a seemingly dead thread, but I experienced the same problem as above, deploying to Heroku. @mdaguete your config.rb fix worked to fix the error.

Thank you.

**Comment by pzi** [comment]
Hi!

I am experiencing the same error during the `middleman build` step of my middleman template on travis-ci: https://travis-ci.org/pzi/middleman-basis#L311

So I'd give this a +1 :neutral_face: 

Thanks for your gem/work!

Cheers,
Patrik

**Comment by Aupajo** [comment]
Hi @pzi,

You can try @mdaguete's approach above. I can't build this into my gem, because modifying the global file watcher ignore list seems like a Bad Idea™ for any library to do. What I probably need to do is reach out to the Middleman team and see if there's any new way around this/anything planned for a future release.

**Comment by pzi** [comment]
Hey @Aupajo,

Yeah, I figured it would be a Bad Idea™ :) I even felt bad for adding it to my `config.rb` and making Travis-CI work more by outputting all those extra/unnecessary lines of code.

I wonder why it's not blowing up when building the site locally. I am using the same ruby version and gems. It only fails on Travis-CI.

Thanks for your help though. Hope that's gonna work out fine!

**Comment by Aupajo** [comment]
I think it's to do with the way Heroku bundles gems into `vendor/`.

**Comment by pzi** [comment]
Hmm, I am not using Heroku though? I just use Travis-CI to run `bundle exec middleman build --verbose` to make sure it all works. Not deploying it to anywhere :|

**Comment by Aupajo** [comment]
Oh, sorry, mis-read your comment. Does Travis do the same? Can you see anything like `bundle install --path ...` anywhere in your Travis log?

**Comment by pzi** [comment]
Hmm, no... this is the bundler line: https://travis-ci.org/pzi/middleman-basis#L89

**Comment by Aupajo** [comment]
That build looks like it's passing?

**Comment by pzi** [comment]
Yeah, because of the ignore in `config.rb` https://github.com/pzi/middleman-basis/blob/develop/config.rb#L91

This is a failed build: https://travis-ci.org/pzi/middleman-basis/builds/54708444

**Comment by Aupajo** [comment]
Ah, thanks! I wonder if there's some Bundler config. Will chat with the Travis folk.

**Comment by pzi** [comment]
Could probably do some trickery with commands:
http://docs.travis-ci.com/user/languages/ruby/#Dependency-Management
and
http://docs.travis-ci.com/user/languages/ruby/#Custom-Bundler-arguments-and-Gemfile-locations

**Comment by Aupajo** [comment]
@pzi From that first page:

> By default, gems are installed into vendor/bundle in your project's root directory.

**Comment by pzi** [comment]
Hmm, so I guess I could do `bundler_args: --path bundle` in my `.travis.yml` :)

Didn't see that. Thanks!

**Comment by Aupajo** [comment]
No problem! Hope it helps the next person.

**Comment by pzi** [comment]
Hmm, ultimately, it has the same effect on ci as before. Except we don't need that hack any more (which feels slightly better).
https://travis-ci.org/pzi/middleman-basis/builds/54819693

**Comment by pzi** [comment]
Hmm, I came across this message: https://groups.google.com/d/msg/travis-ci/XID21_FwgX8/DhW1xU7XlCEJ

And because I have a `Gemfile.lock` in my repo it will do `bundle install --deployment` which causes the problem. So I have a feeling that Heroku would probably do the same given it's a hosting platform.

TL;DR

I felt dirty with `bundler_args: --path bundle` so I overwrote the args that Travis adds, but leave out `--deployment`. Resulting in `bundler_args: --jobs=3 --retry=3` and a properly passing build without hacks.

**Comment by renatocarvalho** [comment]
I'm having this issue as well.

**Comment by Aupajo** [comment]
Have a workaround for the file watcher ignore issue – now running into one where the rendered template is returning `nil`.

Middleman::CoreExtensions::Rendering:

```ruby
def render_individual_file(path, locs={}, opts={}, context=self, &block)
  ...
  body = if opts[:template_body]
    opts.delete(:template_body)
  else
    template_data_for_file(path)
  end
  ...
end
```

`body` is returning `nil`. The `template_data_for_file` method defined in Rendering is not actually being called – the one defined in FrontMatter is, instead.

Middleman::CoreExtensions::FrontMatter:

```ruby
def template_data_for_file(path)
  extensions[:frontmatter].data(path).last
end

def data(path)
  ...
  data, content = frontmatter_and_content(p) # {}, nil
  ...
end

def frontmatter_and_content(path)
  ...
  data = {}
  return [data, nil] if !app.files.exists?(full_path) || ::Middleman::Util.binary?(full_path) # returning here
  ...
end
```

`app.files.exists?(full_path)` is returning `false`.

Returning full circle to the file watcher:

```ruby
def exists?(path)
  p = Pathname(path)
  p = p.relative_path_from(Pathname(@app.root)) unless p.relative?
  @known_paths.include?(p)
end
```

`@known_paths`:

```ruby
@known_paths
=> #<Set: {#<Pathname:config.rb>,
 #<Pathname:source/images/background.png>,
 #<Pathname:source/images/middleman.png>,
 #<Pathname:source/index.html.erb>,
 #<Pathname:source/javascripts/all.js>,
 #<Pathname:source/layouts/layout.erb>,
 #<Pathname:source/stylesheets/all.css>,
 #<Pathname:source/stylesheets/normalize.css>}>
```

Looks like the `reload` trick isn't working any more.

To be continued…

**Comment by hirefrank** [comment]
Having the same problem on CircleCi...

**Comment by hirefrank** [comment]
This worked for me on CicleCI:
```
dependencies:
  override:
    - bundle install --path=gemset: # note ':' here
        timeout: 180 # fail if command has no output for 3 minutes

  # we automatically cache and restore many dependencies between
  # builds. If you need to, you can add custom paths to cache:
  cache_directories:
    - "gemset"   # relative to the build directory
    #- "~/custom_2" # relative to the user's home directory
```

**Comment by Aupajo** [comment]
Still tracking this issue on the Middleman repo: https://github.com/middleman/middleman/issues/1217

At this point, it looks like it won't be an issue in 4.0, but I'm not overly hopeful that it can be easily resolved in 3.X.

**Comment by Aupajo** [comment]
As far as I can tell, this issue has now been addressed and fixed in Middleman 4.

If you're on Middleman 4, you can update your project by altering your `Gemfile` as follows and then running `bundle`:

```ruby
gem 'middleman-search_engine_sitemap', '~> 1.4.0'
```

You should no longer need the `--path` or `file_watcher_ignore` workarounds. 

Let me know if you run into any issues with the new version!
